### ISSUE TITLE
Difference in behaviour with mapped types between concrete and generic types.

### ISSUE THREAD

**Issue by DylanRJohnston** [Microsoft/TypeScript#19509]
Trying to give better types to the redux library. Came across this issue. The issue appears to be the reindexing via `keyof A` at the end.

**TypeScript Version:**  2.7.0-dev.201xxxxx

**Code**

```ts
type Actions = {
  INCREMENT: { x: number }
  DECREMENT: { y: string }
}

type Concrete = {[K in keyof Actions]: {type: K} & Actions[K] }[keyof Actions]

type Generic<A> = {[K in keyof A]: {type: K} & A[K] }[keyof A]

type GenericConcrete = Generic<Actions>

```

**Expected behavior:**
Concrete and GenericConcrete have the same type

**Actual behavior:**
Concrete has the correct type
```ts
type Foo = ({
    type: "INCREMENT";
} & {
    x: number;
}) | ({
    type: "DECREMENT";
} & {
    y: string;
})
```

GenericConcrete has the wrong type

```
type Foo = ({
    type: "INCREMENT" | "DECREMENT";
} & {
    x: number;
}) | ({
    type: "INCREMENT" | "DECREMENT";
} & {
    y: string;
})
```
Unless I'm missing some kind of constraint on A in the Generic case that Typescript is able to use in the concrete case to work out the types correctly?

**Comment by ahejlsberg** [comment]
I believe this was already fixed by #18042 and I can't reproduce this with typescript@next. Can you verify you're  using the latest compiler?

**Comment by DylanRJohnston** [comment]
Yep, I was on an older version of typescript@next. Sorry for the duplicate ticket.
